ext/vulnsrc/rocky/rocky.go

```diff
@@ -0,0 +1,201 @@
+)
+
+const (
+	url        = "https://errata.rockylinux.org/api/advisories"
+	linkFormat = "https://errata.rockylinux.org/%s"
+	updateFlag = "rockyUpdater"
+)
+
+var (
+	rpmRegexp        = regexp.MustCompile(`(?P<name>.*)-(?P<version>[^\-]+\-[^\-]+)\.(?P<basearch>[^.]+)\.rpm$`)
+	rpmReleaseRegexp = regexp.MustCompile(`.*\.el(?P<version>[\d]+)`)
+)
+
+type jsonData struct {
+	Advisories []jsonAdvisory `json"advisories"`
+}
+
+type jsonAdvisory struct {
+	Type             string   `json:"type"`
+	Name             string   `json:"name"`
+	Description      string   `json:"description"`
+	AffectedProducts []string `json:"affectedProducts"`
+	Severity         string   `json:"severity"`
+	CVEs             []string `json:"cves"`
+	RPMs             []string `json:"rpms"`
+}
+
+type updater struct{}
+
+func init() {
+	vulnsrc.RegisterUpdater("rocky", &updater{})
+}
+
+func (u *updater) Update(datastore vulnsrc.DataStore) (resp vulnsrc.UpdateResponse, err error) {
+	log.WithField("package", "Rocky").Info("Start fetching vulnerabilites")
+
+	// Download JSON
+	r, err := httputil.GetWithUserAgent(url)
+	if err != nil {
+		log.WithError(err).Error("could now download Rocky's update")
+		return resp, commonerr.ErrCouldNotDownload
+	}
+	defer r.Body.Close()
+
+	if !httputil.Status2xx(r) {
+		log.WithField("StatusCode", r.StatusCode).Error("Failed to update Rocky")
+		return resp, commonerr.ErrCouldNotDownload
+	}
+
+	// Get the SHA-1 of the latest JSON data
+	latestHash, err := datastore.GetKeyValue(updateFlag)
+	if err != nil {
+		return resp, err
+	}
+
+	// Parse JSON
+	resp, err = buildResponse(r.Body, latestHash)
+	if err != nil {
+		return resp, err
+	}
+
+	return resp, nil
+}
+
+func (u *updater) Clean() {}
+
+func buildResponse(jsonReader io.Reader, latestKnownHash string) (resp vulnsrc.UpdateResponse, err error) {
+	hash := latestKnownHash
+
+	jsonSHA := sha1.New()
+	teedJSONReader := io.TeeReader(jsonReader, jsonSHA)
+
+	// Unmarshal JSON
+	var data jsonData
+	err = json.NewDecoder(teedJSONReader).Decode(&data)
+	if err != nil {
+		log.WithError(err).Error("could not unmarshal Rocky's JSON")
+		return resp, commonerr.ErrCouldNotParse
+	}
+
+	// Calculate the hash and skip updating if the hash has been seen before.
+	hash = hex.EncodeToString(jsonSHA.Sum(nil))
+	if latestKnownHash == hash {
+		log.WithField("package", "Rocky").Debug("no update")
+		return resp, nil
+	}
+
+	// Extract vulnerability data from Rocky's JSON schema.
+	resp.Vulnerabilities = parseRockyJSON(&data)
+
+	return resp, nil
+}
+
+func parseRockyJSON(data *jsonData) (vulnerabilities []database.Vulnerability) {
+	mvulnerabilities := make(map[string]*database.Vulnerability)
+
+	for _, advisory := range data.Advisories {
+		if strings.ToLower(advisory.Type) != "security" {
+			continue
+		}
+
+		vulnName := advisory.Name
+		vulnerability, vulnerabilityAlreadyExists := mvulnerabilities[vulnName]
+
+		if !vulnerabilityAlreadyExists {
+			var subCVEs []string
```

Comment:
Let's pre-allocate the slice here since it's likely we don't filter out much, if any, I think

Reply:
Done